File: 프로그래머스/2/389480. 완전범죄/완전범죄.py
from copy import deepcopy
def solution(info, n, m):
    answer = 10**6
    # DP[a점수][b점수]=True/False
    DP = [[False for _ in range(140)] for _ in range(140)]
    # init
    DP[0][0]=True
    for i in range(len(info)):
        new_DP = [[0 for _ in range(140)] for _ in range(140)]
        check = 0
        for a in range(140):
            for b in range(140):
                if DP[a][b] :
                    if a+info[i][0] < n:
                        new_DP[a+info[i][0]][b] = True
                    if b+info[i][1] < m:
                        new_DP[a][b+info[i][1]] = True
        DP = deepcopy(new_DP)
        
    for i in range(140):
        for j in range(140):
            if DP[i][j] == True:
                answer = min(answer, i)
    if answer == 10**6: return -1
    return answer

File: 프로그래머스/2/389480. 완전범죄/완전범죄.java

class Solution {
    public int solution(int[][] info, int n, int m) {
        int answer = (int)1e6;
        // DP[a점수][b점수]=True/False
        boolean[][] DP = new boolean[140][140];
        // System.out.println(DP);
        // 모두 False로 초기화
        for (int i=0; i<140; i++){
            for (int j=0; j<140; j++){
                DP[i][j] = false;
            }
        }
        DP[0][0]=true;
        for (int i=0; i<info.length ; i++){
            boolean[][] new_DP = new boolean[140][140];
            boolean check = false;
            for (int a=0; a<140; a++){
                for (int b=0; b<140; b++){
                    if (DP[a][b]){
                        if (a+info[i][0] < n){
                            new_DP[a+info[i][0]][b]=true;
                        }
                        if (b+info[i][1] < m){
                            new_DP[a][b+info[i][1]] = true;
                        }
                        
                    }
                }
            }
            // DP를 new_DP로
            for (int a=0; a<140; a++){
                for (int b=0; b<140; b++){
                    DP[a][b] = new_DP[a][b];
                }
            }
        }
        for (int x=0; x<140; x++){
                for (int y=0; y<140; y++){
                    if (DP[x][y]) {
                        answer = Math.min(answer, x);
                    }
                }
            }
        if (answer == (int)1e6){
            return -1;
        }
        return answer;
    }
}